File: train-enwik8.py
#!/usr/bin/env python

import time
import zipfile
from pathlib import Path

import requests
import typer

from quicktok import BasicTokenizer


def main(
    data_dir_path: Path = Path("data"),
    model_file_path: Path = Path("models/enwik8.model"),
    max_bytes: int | None = None,
    num_threads: int = 1,
    vocab_size: int = 2048,
    save_vocab: bool = False,
    enwik8_url: str = "http://mattmahoney.net/dc/enwik8.zip",
):
    zip_file_path = data_dir_path / "enwik8.zip"
    data_file_path = data_dir_path / "enwik8"

    if not zip_file_path.exists():
        response = requests.get(enwik8_url)
        zip_file_path.write_bytes(response.content)

    if not data_file_path.exists():
        with zipfile.ZipFile(zip_file_path, "r") as zip_ref:
            zip_ref.extractall(data_dir_path)

    data = data_file_path.read_bytes()

    if max_bytes:
        data = data[:max_bytes]
        print(f"Training on the first {len(data)} bytes")

    text = data.decode("utf-8", errors="replace")
    model_file_path.parent.mkdir(parents=True, exist_ok=True)
    vocab_file_path = str(model_file_path.with_suffix(".vocab")) if save_vocab else None
    t0 = time.time()
    tokenizer = BasicTokenizer(num_threads)
    tokenizer.train(text, vocab_size)
    tokenizer.save(str(model_file_path), vocab_file_path)
    t1 = time.time()
    print(f"Training took {t1 - t0:.2f} seconds")


if __name__ == "__main__":
    typer.run(main)

File: examples/src/bin/train-enwik8.rs
use std::{
    fs::{self, File},
    io::BufReader,
    path::PathBuf,
    time::Instant,
};

use clap::Parser;
use quicktok::{BasicTokenizer, Tokenizer};
use zip::ZipArchive;

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {
    #[clap(long, default_value = "data")]
    data_dir_path: PathBuf,

    #[clap(long, default_value = "models/enwik8.model")]
    model_file_path: PathBuf,

    #[clap(long, default_value_t = 1)]
    num_threads: usize,

    #[clap(long, default_value_t = 2048)]
    vocab_size: usize,

    #[clap(long)]
    save_vocab: bool,

    #[clap(long, default_value = "http://mattmahoney.net/dc/enwik8.zip")]
    enwik8_url: String,
}

fn main() {
    let args = Args::parse();
    let zip_file_path = args.data_dir_path.join("enwik8.zip");
    let data_file_path = args.data_dir_path.join("enwik8");

    if !zip_file_path.exists() {
        let response =
            reqwest::blocking::get(args.enwik8_url).expect("failed to download enwik8 Zip file");
        fs::write(
            &zip_file_path,
            response.bytes().expect("failed to read response"),
        )
        .expect("failed to write zip file");
    }

    if !data_file_path.exists() {
        let zip_file = File::open(&zip_file_path).expect("failed to open Zip file");
        let mut archive =
            ZipArchive::new(BufReader::new(zip_file)).expect("failed to read Zip file");
        archive
            .extract(&args.data_dir_path)
            .expect("failed to extract Zip file");
    }

    let text = fs::read_to_string(&data_file_path).expect("failed to read data file");
    let output_dir_path = args.model_file_path.parent().unwrap();
    fs::create_dir_all(output_dir_path).expect("failed to create output directory");
    let vocab_file_path = args.model_file_path.with_extension("vocab");
    let start_time = Instant::now();
    let mut tokenizer = BasicTokenizer::new(args.num_threads);
    tokenizer.train(&text, args.vocab_size);
    tokenizer
        .save(
            args.model_file_path.to_str().unwrap(),
            if args.save_vocab {
                Some(vocab_file_path.to_str().unwrap())
            } else {
                None
            },
        )
        .expect("failed to save tokenizer model and vocabulary");
    let duration = start_time.elapsed();
    println!("Training took {:.2} seconds", duration.as_secs_f32());
}
